### PR TITLE
enhancement: add `forc index revert` command

### LINKED ISSUE #464: Add `revert` command to `forc index`
In the event that a new asset deployment breaks a previously working index, there should be a way to quickly revert back to the previous version of the index so that indexing is interrupted as little as possible. This would need a new command to be added to `forc index` as well as a new endpoint in the API server.

### PR DESCRIPTION
**Pull request by rvmelkonian**
Closes #464

## Changelog 
• creates new `forc index revert command` and associated files for the new command to work.  
• updates docs for the new command
• adds new route for the revert op
• adds new queries method `penultimate_index_id_for`
• adds new match arm for `create_service_task()` inside `service.rs`.

## Testing
1. Start the service with a fresh database and using the block explorer example: `cargo run --bin fuel-indexer -- run --fuel-node-host node-beta-2.fuel.network --fuel-node-port 80`
2. Deploy the block explorer example with a log line.
3. Deploy the block explorer example a second time with a different log line.
4. Stop the service.
5. Immediately start it again.
6. Revert the index WASM using the explorer manifest.

 ```
 ⬅️  Revertingindex 'my_project.test_index' at http://127.0.0.1:29987/api/index/my_project/test_index

{
  "success": "true"
}

✅ Successfully reverted index 'my_project.test_index' at http://127.0.0.1:29987/api/index/my_project/test_index
```
If you don't see this make sure to rebuild the binary in step 1.

7. Trigger some events manually 
8. Check in the DB that the penultimate id of `hello-index` (id 1) is indexing the events, not id 2

## Notes
I noticed that in all the other plugin commands, the service is interacted with via the api, which I think is a good and clean approach. As there was no api route for reverting an index, I created a new one. This then triggers the `create_servce_task()` and the new `ServiceRequest::IndexRevert` is matched and executed.

### CODE REVIEW

File: docs/README.md
Comment (ra0x3) [inline docs/README.md:79]: Per the discussion in https://github.com/FuelLabs/fuel-indexer/issues/590, I don't think `beta-3` is ready yet?

Comment (ra0x3) [inline docs/README.md:79]: Per the discussion in https://github.com/FuelLabs/fuel-indexer/issues/590, I don't think `beta-3` is ready yet?

File: docs/src/reference-guide/plugins/forc-index/index.md
Comment (ra0x3) [inline docs/src/reference-guide/plugins/forc-index/index.md:25]: I should probably clarify some terminology at our next team sync as I think I started this trend of calling indexers indices 😅 
```suggestion
    revert    Revert a running indexer to its last previous version
```

File: packages/fuel-indexer-api-server/src/uses.rs
Comment (ra0x3) [inline packages/fuel-indexer-api-server/src/uses.rs:186]: TIL [penultimate](https://www.merriam-webster.com/dictionary/penultimate). Love it.

File: packages/fuel-indexer-api-server/src/uses.rs
Comment (ra0x3) [inline packages/fuel-indexer-api-server/src/uses.rs:183]: Since `penultimate_index_id_for` is just a `SELECT` I don't think we need a transaction?

**Review by ra0x3 — CHANGES_REQUESTED** [review]
@rvmelkonian 
- Awesome work on this so far, looks great! 🥳 
- Left a few small pieces of feedback
- Will test once those ☝🏽 are addressed

File: packages/fuel-indexer-api-server/src/uses.rs
Comment (rvmelkonian) [reply packages/fuel-indexer-api-server/src/uses.rs:183]: Maybe I'm misunderstanding the code, but don't we need to fire a `tx.send()` so that `rx.try_recv()` in `service.rs` will receive the value?

Comment (ra0x3) [reply packages/fuel-indexer-api-server/src/uses.rs:183]: @rvmelkonian 
- You're correct that we're going to send the results of `penultimate_index_id_for` to the indexer service via `tx.send`
- However, since `penultimate_index_id_for` isn't a mutation query (e.g., an `INSERT` or `DELETE`) then we don't really need a transaction
  - The other use cases in `uses.rs` that use transactions are queries that mutate the state of the database

Comment (rvmelkonian) [reply packages/fuel-indexer-api-server/src/uses.rs:183]: got it, yes we are not mutating here. 
So we need `tx.send` but not this:
```
    let _ = queries::start_transaction(&mut conn).await?;

    if let Err(_e) =
        queries::penultimate_index_id_for(&mut conn, &namespace, &identifier).await
    {
        queries::revert_transaction(&mut conn).await?;
    } else {
        queries::commit_transaction(&mut conn).await?;
    }
    ```

File: packages/fuel-indexer-database/postgres/src/lib.rs
Comment (deekerno) [inline packages/fuel-indexer-database/postgres/src/lib.rs:625]: Interestingly, I get this error message whenever I do the revert operation:
```
ERROR fuel_indexer::service: 328: Failed to find Index(fuel_examples.explorer_index): error returned from database: LIMIT #,# syntax is not supported
```

File: plugins/forc-index/src/ops/forc_index_revert.rs
Comment (deekerno) [inline plugins/forc-index/src/ops/forc_index_revert.rs:22]: Can we remove this as the user has to pass the manifest path in for the command to work?

**Review by deekerno — CHANGES_REQUESTED** [review]
Looking great, I think it's almost there! I left a couple of comments.

Additionally, I think we should consider removing the reverted index from the database; if you restart the service after reverting an index, the previously reverted index will be the one executed.

Reproduction steps:
- Deploy an index with a log message included, e.g. "This is the first index." You should see this in the service logs.
- Deploy a second index with a different log message, e.g. "This is the second index." Verify that it's in the logs.
- Revert the second index.
- Stop the index and start it again.
- You should see that the second index's message is the one that appears.

File: packages/fuel-indexer-database/postgres/src/lib.rs
Comment (rvmelkonian) [reply packages/fuel-indexer-database/postgres/src/lib.rs:625]: hmm interesting, I haven't seen that error on my side. Did you run migrations before testing? What postgres version are you on?

**Comment by rvmelkonian** [comment]
@deekerno Yes I had also thought about removing the reverted index from the DB. I will update this with that logic, I think it makes sense and of course removes the bug you've mentioned.

**Comment by rvmelkonian** [comment]
@deekerno I've added the `queries::remove_index` method *after* the reversion happens, inside service.rs. I did this so that the delete mutation will only happen if `queries::penultimate_index_id_for` does not return an error. The only downside I can think of this is that we don't have the safety of doing the DELETE within a transaction. Lmk what you think about this. 

As it stands now, after the revert command, the reverted index would be the latest index in the DB.

File: packages/fuel-indexer/src/service.rs
Comment (deekerno) [inline packages/fuel-indexer/src/service.rs:342]: So `remove_index` removes the _entire_ index, as it deletes any rows that match to the `index_id` for the namespace and identifier. In contrast, I'd expect a removal of the reverted index to remove the most recent `version` of the WASM asset for said `index_id`. My apologies if I caused a misunderstanding when suggesting it earlier.

This would require a separate function to be created or for `remove_index` to take at least another argument to instruct what should be deleted.

@ra0x3: Thoughts?

File: packages/fuel-indexer-database/postgres/src/lib.rs
Comment (deekerno) [reply packages/fuel-indexer-database/postgres/src/lib.rs:627]: PostgreSQL 14, the one that's installed through `brew install postgresql`. And yep, I've run the migrations fresh on each test.

It looks like you can use `LIMIT X OFFSET Y` to skip an arbitrary number of rows: https://www.postgresql.org/docs/current/queries-limit.html; testing it locally worked for me.

**Review by deekerno — CHANGES_REQUESTED** [review]
@rvmelkonian: Left another comment concerning the removal of the reverted index. Once that's addressed, I'll re-test and approve.

Regarding not having a transaction, is there any reason that we couldn't start one in the `Ok(id)` match arm of `queries::penultimate_index_id_for(..)`? I think we should be able to start one with the same connection by doing ```let _ = queries::start_transaction(&mut conn).await?;```

**Comment by rvmelkonian** [comment]
@deekerno I've tested with you steps above https://github.com/FuelLabs/fuel-indexer/pull/609#pullrequestreview-1322503833 and it's working as expected. However, note that after reverting, the indexer has to be restarted for the penultimate wasm asset to be the active index. 

After doing this revert I'm able to successfully continue deploying other indexes.

**Review by deekerno — DISMISSED** [review]
LGTM! Thanks for adding this!

Eventually, we should work on our API errors across the product to ensure that users have a better idea of what's going on when things go wrong. But this is definitely good for now.

File: packages/fuel-indexer-api-server/src/api.rs
Comment (ra0x3) [inline packages/fuel-indexer-api-server/src/api.rs:11]: 👌🏽

**Review by ra0x3 — CHANGES_REQUESTED** [review]
@rvmelkonian 
- The logic looks fine
- Left some minor comments and one comment about moving some of the indexer swap logic out of the service module and into the use case module

**Review by ra0x3 — COMMENTED** [review]
- [ ] Deployed the block explorer index
- [ ] Made changes to the example then redeployed
- [ ] Confirmed that the 2nd index replaced the 1st
- [ ] Ran `forc-index revert --path /path/to/indexer` and saw that the latest version was removed while 
- [ ] Confirmed that the 1st indexer came back while the 2nd was removed

@rvmelkonian This works, will approve after the last small feedback is addressed

File: packages/fuel-indexer/src/service.rs
Comment (rvmelkonian) [reply packages/fuel-indexer/src/service.rs]: @ra0x3 this has been added in now.

**Review by ra0x3 — APPROVED** [review]
:shipit:

**Pull request merged**
